### DISCUSSION TITLE
Does the name should be checked in X.509 path establishment?

### DISCUSSION THREAD

**Discussion by x509-name-testing** [strongswan/strongswan#1900] [General]
Hi there,

I am writing to ask a question about the PKI implementation in this library. When I tested the path-establishing process, I found that the "name" checking seems not critical. Here is an example output when using the tool of `StrongSwan`'s `pki`

```
> pki --verify --cacert ../tmp/ca.pem --in ../tmp/ee.pem 
  file coded in unknown format, discarded
building CRED_CERTIFICATE - X509 failed, tried 3 builders
parsing CA certificate from '../tmp/ca.pem' failed
  using trusted certificate "CN=aaaa"
no issuer certificate found for "CN=aaaa"
  issuer is "CN=bbbb"
certificate trusted, lifetimes valid
```

The used certificates (ca, ee) are listed here. Ca cert:
```
-----BEGIN CERTIFICATE-----
MIICyjCCAbKgAwIBAgICECAwDQYJKoZIhvcNAQELBQAwDzENMAsGA1UEAxMEQUFB
YTAeFw0xNTAyMDcxNzI0MDBaFw0yNDAyMTQwNjI2NTNaMA8xDTALBgNVBAMTBEFB
QWEwggEgMA0GCSqGSIb3DQEBAQUAA4IBDQAwggEIAoIBAQDMdoaoVMhsqfd3k2KU
Kcod5LEQdAlXewJQOzB9OMmnErAWsOcEsKkvYLIDNczRWC0f4ssQh59leBvfkW3P
brpEgyMaSp2OKuzSDKPEs80MvWrVfty7HLtDYbfmb1tTIVzMHyo9lL4wMRMuqGTZ
3fwWZKSJ/BwDMrTnqXiV+lTvF3aD2zP/qRD8vJu2vCEb+6nyidC6tZlmMzEn+hBV
aHEUPqtI0V5YPexpFGi37QshH3Fu1/fKBxXxd/5ZAHHrPtsCcEapVKaY2ECSQmmY
KCLENvg1KPCcumqwgojjvtE8gjs+hYBLu2g3ppIYlnndpuPDfOJL0kQun0GG8jkT
Y5QDAgEDozIwMDAPBgNVHRMBAf8EBTADAQH/MB0GA1UdDgQWBBRr+aQtpek5iagk
WHmHEfxvB5HvpjANBgkqhkiG9w0BAQsFAAOCAQEAV8Upn8T68Pieov34vVDt/3vJ
BUMJym5qssmshI/JJVVJR1dauzqdgRM0Gm9Ib7LB2AH9yh6lA7aYaqZxz5/rl+98
z8+PqyBI5D3mdlzc+pjmsCstuReibWWn84OjDoEVCG2xbgUE1+SzfNs6jX4YJwMA
Cg43KzxhceXuYvYtzYZ07NQ8x5qfbWRu9uDEzQEOWrRHn73gixYzD+gXX0MbsZty
mIibkU5nMEmOWd4tMePe0WtmltImLuJ+v00xMA4fdgW9N14uIx5OvrG49G5OlZjA
+uXp03wBwNOkjZQ3FI7W3GNzkobgbvUkeoAL2qY0kNAls62Eqe9/Pi6IRJUWkg==
-----END CERTIFICATE-----
```
Ee cert:
```
-----BEGIN CERTIFICATE-----
MIIC+TCCAeGgAwIBAgICECEwDQYJKoZIhvcNAQELBQAwDzENMAsGA1UEAxMEYmJi
YjAeFw0xNjAyMDcxNzI0MDBaFw0yNDAyMTQwNjI2NTNaMA8xDTALBgNVBAMTBGFh
YWEwggEgMA0GCSqGSIb3DQEBAQUAA4IBDQAwggEIAoIBAQCoil4mIxsx0zcacLLs
P3TUtETjeqXA9aqXJpoE/9q+5QkDmD21vwEsmgo6+7w8576DXLNw6Fzj0YPDlAjN
Gofl4FucXG6wfeJYbMO1yJ0R8V2WDWYeVn+PWael4cXngUwJnV6W8JrCi3DVq3lY
XbdYqv11UqpLpyVodlkA7ngrkcZZkZk4PqF2w/Uja+YH6mMcl0nvoP79E8mpn8IL
5oeSW8z1QpVKpG1kun3OywTQ+Ofj2nVg04tqZPx4ViFpWuinj/uPguOuNqKTZpLL
gqO+hACG3H5tU3eEF7lVQw3xFh/VQ3WZZhlS0KxfdK2ykBVQBHRD32w10P0yN7ON
9eUJAgEDo2EwXzAMBgNVHRMBAf8EAjAAMA8GA1UdEQQIMAaCBGFhYWEwHQYDVR0O
BBYEFJJqHlI6GlefyYKazsjAqVGdL8dyMB8GA1UdIwQYMBaAFGv5pC2l6TmJqCRY
eYcR/G8Hke+mMA0GCSqGSIb3DQEBCwUAA4IBAQBNm0Y2P5+lTsQJf8V8jFbASeZ7
cRBMdDA2pJtZN+bF8/3q2E/BTWRjYubsDMeYZx2Ltv0b4fqPGN6bMcH5u4HxUaZX
4Rkxfu+ijANPTApfqNZw3zFPSCR1jPYiPYmVxQyG0Zz4e/p+WrOadT4aeJUvaaso
+lwvPc5te3Ru8yMn00IqNMtO5GBCTB5ibe7sJxXa0Uk/dJQeZfw3kfaBLOc+OpJs
J2uD9yDXpyUcqRt45Jng5DDYY8FW2gHNV1ICyij6Pie1q/EOP2Sqb2JZc2VAztYi
/NtSzUT5heKJedygtJwoAkjMUbbSJ9wT/bkLyDBf137oK8jNVJIVlWkr3Ou2
-----END CERTIFICATE-----
```

Does it mean the `name` checking, for example, common name, is not necessary in `StrongSwan`? Because the related specifications said the name checking should not be omitted, so I am here for enlightenment.

Looking forward to your reply. Thanks.

Regards.

**Comment by tobiasbrunner** [comment]
`pki --verify` does no identity checks for the end-entity certificate. It just checks the certificate chain.

**Reply by x509-name-testing** [reply]
Thanks for your reply.

I am using the `pki --verify` to verify the certificate chain, which I showed previously. I think the chain should not be verified as "trusted" because the ca-cert's subject name is not one-to-one matched to the ee-cert's issuer name. The ca-cert's subject field should be `CN=AAAa` while the ee-cert's issuer field should be `CN=bbbb`, which is obviously not the same. But because the crypto-related attributes are validated, for example, signature, so the `pki --verify` prints out the "certificate trusted." May I know if my understanding is correct? 

I got confused because, from the specifications like RFC 5280, the `name` components should be one-to-one matched. I occasionally find it is different from the RFC's description. So I am writing for more details, I will much appreciate it if you can give me some information. 

I am looking forward to your reply. Many thanks.

Regards.

**Reply by tobiasbrunner** [reply]
I see. If an authorityKeyIdentifier is encoded, that's used to build the chain (matching it against subjectKeyIdentifier of the issuer certificate). The issuer field is only used as a fallback.
